business/apps.go

```diff
@@ -279,3 +289,33 @@ func fetchNamespaceApps(layer *Layer, namespace string, appName string) (namespa
 
 	return allEntities, nil
 }
+
+type appServiceWithTracing struct {
+	AppService
+}
+
+func (in *appServiceWithTracing) GetAppList(ctx context.Context, namespace string, linkIstioResources bool) (models.AppList, error) {
+	var span trace.Span
+	ctx, span = otel.Tracer(observability.TracerName()).Start(ctx, "GetAppList",
```

Comment:
Suggestion: as we have an "observability" package, I wonder if it looks good to create a helper to propagate the trace/span attributes from the "observability" package instead to propagate the import of the "go.opentelemetry.io" packages almost on every business package.

Something similar as the "kiali/prometheus/internalmetrics" implemented just to pass the "context" and attributes and make the "go.opentelemetry.io" part of the implementation detail of the "observability" package.

Reply:
Another idea is that if you move the tracing logic into an external "observability" package, perhaps you don't need to deal with two service implementations. The tracing method can always be invoked (I'm just thinking of the same design as the internal metrics). The tracing logic is managed internally. "observability" logic is always invoked, but only spans are added when the tracing service is enabled.

Let me know if that idea is clear, perhaps I could prototype some example, but basically, it would look similar to the usage of the internal metric we had in the past (now we have that service placed at handler level, but in the past, it was located in the same place you are proposing the tracing instrumentation).

Reply:
> Suggestion: as we have an "observability" package, I wonder if it looks good to create a helper to propagate the trace/span attributes from the "observability" package instead to propagate the import of the "go.opentelemetry.io" packages almost on every business package.

Sure I can include all those imports and re-export them as part of the observability package.

> Another idea is that if you move the tracing logic into an external "observability" package, perhaps you don't need to deal with two service implementations. The tracing method can always be invoked (I'm just thinking of the same design as the internal metrics). The tracing logic is managed internally. "observability" logic is always invoked, but only spans are added when the tracing service is enabled.

I'll play around with that a bit. This is more of the approach I originally went with having checks for `config.Observability.Tracing.Enabled` before starting the span.

Reply:
> I'll play around with that a bit. This is more of the approach I originally went with having checks for config.Observability.Tracing.Enabled before starting the span.

From the instrumented code, the call to the observability package could be always present and inside the implementation, it can be checked if the service is enabled/disabled. It's a tradeoff but the cost would be just an empty call instead to bring the check on any business package.

But that direction looks +1.

Reply:
@lucasponce I implemented a small change exposing the span start as part of the observability package: https://github.com/kiali/kiali/pull/4532/commits/a4f3c2f9c9cf38bf347d51ba87308f3d21904ba8. The API is probably a bit rough but I'm not convinced this adds much value over just putting checks in each of the business methods. It basically just abstracts away the `config.Get().Observability.Tracing.Enabled` check.

The problem comes in needing to call `defer span.End()` within the business function itself. Which means you either need to check `span` for nil if you return `span` or go with some kind of cleanup function. Either way it's a bit awkward to use. Open to suggestions here on if this can be made better.

Otherwise I'm fine with either adding the span management directly to the business methods or else keeping the tracing logic in a separate implementation.